src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,158 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> next;
+        private Node<T> previous;
+
+        public Node(Node<T> previous, T value, Node<T> next) {
+            this.previous = previous;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            head = new Node<>(null, value, null);
+            tail = head;
+        } else {
+            Node<T> newNode = new Node<>(tail, value, null);
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkForSuitableIndex(index);
+        Node<T> newNode = new Node<>(null, value, null);
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            addFirst(newNode);
+        } else {
+            Node<T> currentNode = getNodeByIndex(index);
+            newNode.next = currentNode;
+            newNode.previous = currentNode.previous;
+            currentNode.previous.next = newNode;
+            currentNode.previous = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null) {
+            throw new RuntimeException("The list is null!");
+        }
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
```

Comment:
I think you can make index check in getNodeByIndex , it will optimize your code and you won`t call checkIndex many times, just once when you call getNodeByIndex

Reply:
> I think you can make index check in getNodeByIndex , it will optimize your code and you won`t call checkIndex many times, just once when you call getNodeByIndex
i think it is not a good point bec in my add method there are checkIndex method and getNodeByIndex. So in this case i will check index 2 times where the second is unnecessary. So it will not work properly(  And create same methods but one without check is stupid)

Reply:
and thx for comments)